contrib/sbtc/token-key-admin/contracts/sbtc-token-key-admin.clar

```diff
@@ -0,0 +1,200 @@
+
+;; token definitions
+;; 
+(define-fungible-token sbtc u21000000)
+
+;; constants
+;;
+(define-constant err-invalid-caller u1)
+(define-constant err-invalid-signer-id u2)
+(define-constant err-not-token-owner u3)
+(define-constant err-trading-halted u4)
+
+;; data vars
+;;
+(define-data-var contract-owner principal tx-sender)
+(define-data-var coordinator (optional {addr: principal, key: (buff 33)}) none)
+(define-data-var num-keys uint u4000)
+(define-data-var num-parties uint u4000)
+(define-data-var threshold uint u2800)
+(define-data-var bitcoin-wallet-address (optional (string-ascii 72)) none)
+(define-data-var trading-halted bool false)
+
+;; data maps
+;;
+(define-map signers uint {addr: principal, key: (buff 33)})
+
+;; public functions
+;;
+(define-public (set-contract-owner (owner principal))
+    (begin
+        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (ok (var-set contract-owner owner))
+    )
+)
+
+(define-public (set-coordinator-data (data {addr: principal, key: (buff 33)}))
+    (begin
+        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (ok (var-set coordinator (some data)))
+    )
+)
+
+(define-public (set-num-keys (n uint))
+    (begin
+        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (ok (var-set num-keys n))
+    )
+)
+
+(define-public (set-num-parties (n uint))
+    (begin
+        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (ok (var-set num-parties n))
+    )
+)
+
+(define-public (set-threshold (n uint))
+    (begin
+        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (ok (var-set threshold n))
+    )
+)
+
+(define-public (set-trading-halted (b bool))
+    (begin
+        (asserts! (is-coordinator) (err err-invalid-caller))
+        (ok (var-set trading-halted b))
+    )
+)
+
+(define-public (set-signer-data (id uint) (data {addr: principal, key: (buff 33)}))
+    (begin
+        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (asserts! (is-valid-signer-id id) (err err-invalid-signer-id))
+        (ok (map-set signers id data))
+    )
+)
+
+(define-public (delete-signer-data (id uint))
+    (begin
+        (asserts! (is-contract-owner) (err err-invalid-caller))
+        (asserts! (is-valid-signer-id id) (err err-invalid-signer-id))
+        (ok (map-delete signers id))
+    )
+)
+
+(define-public (set-bitcoin-wallet-address (addr (string-ascii 72)))
+    (begin
+        (asserts! (is-coordinator) (err err-invalid-caller))
+        (ok (var-set bitcoin-wallet-address (some addr)))
+    )
+)
+
+(define-public (mint! (amount uint) (memo (string-ascii 72)))
+    (begin
+        (asserts! (is-coordinator) (err err-invalid-caller))
+        (ft-mint? sbtc amount tx-sender)
+    )
+)
+
+(define-public (burn! (amount uint) (memo (string-ascii 72)))
+    (begin
+        (asserts! (is-coordinator) (err err-invalid-caller))
+        (ft-burn? sbtc amount tx-sender)
+    )
+)
```

Comment:
These methods should take an additional principal as an argument, since the coordinator will mint & burn tokens for arbitrary stacks principals.

Reply:
I think `memo` is meant to encode the bitcoin ops request txid.

Reply:
Done @netrome

Reply:
See below @igorsyl , I used peg-in/out-txid for mint/burn